### ISSUE TITLE
MATSimGUI → Create Sample Population can deadlock

### ISSUE THREAD

**Issue by kainagel** [matsim-org/matsim-libs#1462]
(MATSim)GUI has a feature to downsample the population.  See the following screen shot.

![image](https://user-images.githubusercontent.com/8777649/115296016-b32afe80-a15a-11eb-86cb-2e4c4f678847.png)

If the population already has routes, then the population reader needs the network.  So in the case I tried today during teaching (scenario/equil/plans100.xml from the matsim-example-project).  It first throws an exception, but then goes on and requests the network.  But the network choice dialogue did not show up.

I tried again tonight, and the dialogue sometimes shows up and sometimes not.  If I put additional log statements between all the executing lines around line 186 in `PopulationSampler`, then it seems to come up much more deterministically.

I ran jstack on a stuck execution; see attached. It states that there is a thread deadlock.  Unfortunately, I do not know enough about multithreading to go further here.  Anybody any idea?

I ran this out of the IDE from MATSimGUI in matsim-example-project, if that makes a difference.

[jstack.log](https://github.com/matsim-org/matsim-libs/files/6338742/jstack.log)

**Comment by michalmac** [comment]
I think the problem is caused by calling the "show dialog" methods from outside the Event Dispatch Thread. I will try to rearrange the code a bit.

**Comment by kainagel** [comment]
Thank you so much!

**Comment by michalmac** [comment]
@kainagel I fixed some Swing threading issues in #1470. Please give it a try (I was not able to reproduce the deadlock).

**Comment by michalmac** [comment]
Adding a short analysis of the log in case we run into similar issues in the future:

The EDT got stuck on processing a `tableRowsInserted` event:
```
"AWT-EventQueue-0" #30 prio=6 os_prio=31 cpu=1115.36ms elapsed=304.88s tid=0x00007f88de2a7000 nid=0xe913 waiting for monitor entry  [0x000070001115b000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at java.awt.Component.invalidate(java.desktop@14.0.1/Component.java:3011)
	- waiting to lock <0x0000000400781f60> (a java.awt.Component$AWTTreeLock)
	at java.awt.Container.invalidate(java.desktop@14.0.1/Container.java:1614)
	at javax.swing.JComponent.revalidate(java.desktop@14.0.1/JComponent.java:4891)
	at javax.swing.JTable.tableRowsInserted(java.desktop@14.0.1/JTable.java:4487)
	at javax.swing.JTable.tableChanged(java.desktop@14.0.1/JTable.java:4412)
	at javax.swing.table.AbstractTableModel.fireTableChanged(java.desktop@14.0.1/AbstractTableModel.java:297)
	at javax.swing.table.AbstractTableModel.fireTableRowsInserted(java.desktop@14.0.1/AbstractTableModel.java:232)
	at com.apple.laf.AquaFileSystemModel$DoChangeContents.run(java.desktop@14.0.1/AquaFileSystemModel.java:477)
	- locked <0x0000000403a87d98> (a java.lang.Object)
...
```
At the same time, the "SampleCreator" thread was operating on GUI (including the same table):
```
"SampleCreator" #51 prio=6 os_prio=31 cpu=516.62ms elapsed=263.75s tid=0x00007f88c8960000 nid=0x17dc3 waiting for monitor entry  [0x00007000113e4000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at com.apple.laf.AquaFileSystemModel.getRowCount(java.desktop@14.0.1/AquaFileSystemModel.java:203)
	- waiting to lock <0x0000000403a87d98> (a java.lang.Object)
	at javax.swing.JTable.getRowCount(java.desktop@14.0.1/JTable.java:2650)
	at javax.swing.plaf.basic.BasicTableUI.createTableSize(java.desktop@14.0.1/BasicTableUI.java:1732)
	at javax.swing.plaf.basic.BasicTableUI.getPreferredSize(java.desktop@14.0.1/BasicTableUI.java:1773)
	at javax.swing.JComponent.getPreferredSize(java.desktop@14.0.1/JComponent.java:1680)
	at javax.swing.ScrollPaneLayout.preferredLayoutSize(java.desktop@14.0.1/ScrollPaneLayout.java:499)
	at java.awt.Container.preferredSize(java.desktop@14.0.1/Container.java:1823)
	- locked <0x0000000400781f60> (a java.awt.Component$AWTTreeLock)
```
So the deadlock was on the locks(monitors): 0x0000000400781f60 and 0x0000000403a87d98.

The actual cause of the problem was opening JFileChooser directly from the "SampleCreator" thread (i.e. outside EDT):
```
	at javax.swing.JFileChooser.showOpenDialog(java.desktop@14.0.1/JFileChooser.java:653)
	at org.matsim.run.gui.PopulationSampler.createSample(PopulationSampler.java:186)
	at org.matsim.run.gui.PopulationSampler$2$1.run(PopulationSampler.java:145)
	at java.lang.Thread.run(java.base@14.0.1/Thread.java:832)
```
